src/toil/batchSystems/slurm.py

```diff
@@ -444,6 +442,37 @@ def prepareSbatch(self,
             if cpu is not None:
                 sbatch_line.append(f'--cpus-per-task={math.ceil(cpu)}')
 
+            if gpus:
+                sbatch_line = sbatch_line[:1] + [f'--gres=gpu:{gpus}'] + sbatch_line[1:]
+
+                lowest_priority_gpus = sorted([i for i in self.boss.partitions if i[1]], key=lambda x: x[4])
+                new_partition_name = lowest_priority_gpus[0][0]  # get name of lowest priority partition that supports gpus
+                if not any(line.startswith("--partition") for line in sbatch_line):
+                    # no partition specified, so specify one
+                    sbatch_line.append(f"--partition={new_partition_name}")
+                else:
+                    # there is a partition specified already, check if the partition has GPUs
+                    for i, line in enumerate(sbatch_line):
+                        if line.startswith("--partition"):
+                            # grab the partition name depending on if it's specified via an "=" or a space
+                            if "=" in line:
+                                separated = False
+                                partition_name = line[len("--partition="):]
+                            else:
+                                separated = True
+                                partition_name = line[i+1]
+
+                            if not any(partition_name == partition for partition in lowest_priority_gpus):
+                                # the specified partition is not compatible, so override with a gpu supported partition
+                                logger.warning(f"Job {jobName} needs {gpus} GPUs, but specified partition {partition_name} is incompatible. Overriding with partition {new_partition_name}.")
+                                # remove old partition from sbatch line
+                                sbatch_line.pop(i)
+                                if separated:
+                                    sbatch_line.pop(i)
+                                # add new partition to sbatch line
+                                sbatch_line.append(f"--partition={new_partition_name}")
+                            break
```

Comment:
I don't think we need or really even want to override the user's `--partition` option. I might be a user who under no circumstances wants my job to run in the `expensive_gpus` partition, and I might put `--partition=normal` explicitly to make sure that can never happen, and I might then have Toil helpfully ignore what I told it to do and drop stuff in the `expensive_gpus` partition anyway, and then I might have to pay a bunch of money to my university's IT department.

If the user explicitly sets a partition in the Slurm arguments, we probably should drop all the jobs in there even if we think that they will never schedule. We might want to warn the user that we think this isn't going to work (or maybe even fail the run?), but we shouldn't just directly override explicit instructions with auto-detected guesses.